Dockerfile

```diff
@@ -0,0 +1,375 @@
+#                        essentials. Instead the ${HOME}/.local folder
+#                        is copied from the build-image - this way we have
+#                        only result of installation and we do not need
+#                        all the build essentials. This makes the image
+#                        nuch smaller.
+#
+ARG AIRFLOW_VERSION="2.0.0.dev0"
+ARG AIRFLOW_ORG="apache"
+ARG AIRFLOW_REPO="airflow"
+ARG AIRFLOW_GIT_REFERENCE="master"
+ARG REQUIREMENTS_GIT_REFERENCE="master"
+ARG WWW_FOLDER="www"
+ARG AIRFLOW_EXTRAS="async,aws,azure,celery,dask,elasticsearch,gcp,kubernetes,mysql,postgres,redis,slack,ssh,statsd,virtualenv"
+
+ARG AIRFLOW_HOME=/opt/airflow
+ARG AIRFLOW_UID="50000"
+ARG AIRFLOW_GID="50000"
+
+ARG PIP_VERSION="19.0.2"
+ARG CASS_DRIVER_BUILD_CONCURRENCY="8"
+
+ARG PYTHON_BASE_IMAGE="python:3.6-slim-buster"
+ARG PYTHON_MAJOR_MINOR_VERSION="3.6"
+
+##############################################################################################
+# This is the build image where we build all dependencies
+##############################################################################################
+FROM ${PYTHON_BASE_IMAGE} as airflow-build-image
+SHELL ["/bin/bash", "-o", "pipefail", "-e", "-u", "-x", "-c"]
+
+LABEL org.apache.airflow.docker=true
+LABEL org.apache.airflow.distro="debian"
+LABEL org.apache.airflow.distro.version="buster"
+LABEL org.apache.airflow.module="airflow"
+LABEL org.apache.airflow.component="airflow"
+LABEL org.apache.airflow.image="airflow-build-image"
+LABEL org.apache.airflow.uid="${AIRFLOW_UID}"
+
+ARG AIRFLOW_VERSION
+ARG AIRFLOW_ORG
+ARG AIRFLOW_REPO
+ARG AIRFLOW_GIT_REFERENCE
+ARG REQUIREMENTS_GIT_REFERENCE
+ARG WWW_FOLDER
+ARG AIRFLOW_EXTRAS
+
+ARG AIRFLOW_HOME
+ARG AIRFLOW_UID
+ARG AIRFLOW_GID
+
+ARG PIP_VERSION
+ARG CASS_DRIVER_BUILD_CONCURRENCY
+
+ARG PYTHON_MAJOR_MINOR_VERSION
+
+ENV PYTHON_BASE_IMAGE=${PYTHON_BASE_IMAGE}
+ENV PYTHON_MAJOR_MINOR_VERSION=${PYTHON_MAJOR_MINOR_VERSION}
+
+ENV AIRFLOW_VERSION=${AIRFLOW_VERSION}
+ENV AIRFLOW_ORG=${AIRFLOW_ORG}
+ENV AIRFLOW_REPO=${AIRFLOW_REPO}
+ENV AIRFLOW_GIT_REFERENCE=${AIRFLOW_GIT_REFERENCE}
+ENV REQUIREMENTS_GIT_REFERENCE=${REQUIREMENTS_GIT_REFERENCE}
+
+ENV AIRFLOW_EXTRAS=${AIRFLOW_EXTRAS}
+
+ENV AIRFLOW_REPO_URL="https://github.com/${AIRFLOW_ORG}/${AIRFLOW_REPO}"
+ENV AIRFLOW_RAW_CONTENT_URL="https://raw.githubusercontent.com/${AIRFLOW_ORG}/${AIRFLOW_REPO}"
+
+ENV PIP_VERSION=${PIP_VERSION}
+ENV CASS_DRIVER_BUILD_CONCURRENCY=${CASS_DRIVER_BUILD_CONCURRENCY}
+
+ARG AIRFLOW_SOURCES="${AIRFLOW_REPO_URL}/archive/${AIRFLOW_GIT_REFERENCE}.tar.gz#egg=apache-airflow"
+ENV AIRFLOW_SOURCES=${AIRFLOW_SOURCES}
+
+ARG CONSTRAINT_REQUIREMENTS="${AIRFLOW_RAW_CONTENT_URL}/${REQUIREMENTS_GIT_REFERENCE}/requirements/requirements-python${PYTHON_MAJOR_MINOR_VERSION}.txt"
+ENV CONSTRAINT_REQUIREMENTS=${CONSTRAINT_REQUIREMENTS}
+
+ARG ENTRYPOINT_FILE="${AIRFLOW_RAW_CONTENT_URL}/${REQUIREMENTS_GIT_REFERENCE}/entrypoint.sh"
```

Comment:
Why is this needed? Isn't `COPY entrypoint.sh /entrypoint` the same and easier to reason about?

Reply:
Not really. We have two ways of  building the image:

1) From local sources (and then we can use entrypoint.sh locally)
2) From remote Github URL. And then we need to make sure that the entrypoint.sh file is the same as the version we are installing the image from. We will be able to use the same Dockerfile to install airflow from both 1.10 and 2.0 and there will be difference between those two entryopoint.sh files ('airflow resetdb` vs `airflow db reset` is one that will be there for sure). So we need to make sure to use the correct one.

I want this Dockerfile to become the source of the "official" DockerHub image (https://docs.docker.com/docker-hub/official_images/). And if I do the above - get everything from released versions in GitHub for the "remote" version of building the image =  then we only need the Dockerfile itself and no additional files to become the "official" docker image. 

This way the Dockerfile becomes a "standalone" Dockerfile.

The idea is that once we release Airflow 1.10.10 you should be able to just take the Dockerfile (and nothing else) put it in an empty folder and run:

``` 
docker build . --build-arg=AIRFLOW_GIT_REFERENCE=1.10.10 --build-arg=AIRFLOW_VERSION=1.10.10
```

and image of Airflow 1.10.10 should be built in a repeatable fashion. 

I managed to achieve exactly this property by adding requirements.txt and downloading everything from github.

Reply:
> This way the Dockerfile becomes a "standalone" Dockerfile.
>
> The idea is that once we release Airflow 1.10.10 you should be able to just take the Dockerfile (and nothing else) put it in an empty folder and run:

This is different to every other project I've ever seen build a docker file. Why do we want this?

Honestly if I was wanting to build an image for version 1.10.10, I'd expect to check out that branch/tag and do `docker build .` or a close variation of it.

Reply:
Official docker images are so you can do `docker pull airflow`? I'm not sure AFS would like us to drop the Apache name there.

Reply:
> Official docker images are so you can do `docker pull airflow`? I'm not sure AFS would like us to drop the Apache name there.

I am definitely planning to pass it through PMC (and ASF if we so decide) before I make a submission. 

And I would actually prefer "apache-airflow" as a name (same as we use for pip package) so it will rather be `docker pull apache-airflow`. I think it's good if you read the guide here about official images here: https://docs.docker.com/docker-hub/official_images/. It was originally intended for languages but then other projects started using it which were more like frameworks (django for one).  It is followed by python, nodejs or django do (see https://hub.docker.com/_/django/) - those are all free foundations-managed projects and the idea with this program is that it certifies that image is maintained by some organization (which it will be in our case). It is prominently displayed, it goes through automated vulnerability scanning for example. I think it's good to have our image as official one.

Just a note - the intention to do that  was clear from the very beginning see already approved AIP-26: https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-26+Production-ready+Airflow+Docker+Image+and+helm+chart - it's point no. 5) in the acceptance criteria.

Voting here: https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-26+Production-ready+Airflow+Docker+Image+and+helm+chart

Reply:
Oh `docker pull apache-airflow` would be fine, yeah.

How quickly can we iterate on our Dockerfile if needed once it makes it in to the "library"

Reply:
It's all on us. We just make a PR with hash of the commit from repo and they approve it.